File: SimPhys_W2/templates/ljfluid.py
from numpy import *
from matplotlib.pyplot import *

def minimum_image(ri, rj):
    global L
    # compute distance
    rij = rj-ri
    # wrap the distance into [-0.5, 0.5]
    rij -= rint(rij/L)*L
    return rij

def compute_lj_force(rij):
    # TODO
    # .
    # .
    # .
    return fij

def compute_lj_potential(rij):
    # TODO
    # .
    # .
    # .
    return pot

def compute_forces(x, L):
    """Compute and return the forces acting onto the particles,
    depending on the positions x."""
    _, N = x.shape
    f = zeros_like(x)
    rijs = empty_like(x)
    for i in range(1,N):
        for j in range(i):
            # distance vector
            rij = minimum_image(x[:,i], x[:,j])
            fij = compute_lj_force(rij)
            f[:,i] -= fij
            f[:,j] += fij
    return f

def compute_energy(x, v, L):
    """Compute and return the total energy of the system with the
    particles at positions x."""
    _, N = x.shape
    E_pot = 0.0
    E_kin = 0.0
    # sum up potential energies
    for i in range(1,N):
        for j in range(i):
            # distance vector
            rij = minimum_image(x[:,i], x[:,j])
            E_pot += compute_lj_potential(rij)
    # sum up kinetic energy
    for i in range(N):
        E_kin += 0.5 * dot(v[:,i],v[:,i])
    return E_pot + E_kin
    
def step_vv(x, v, f, dt):
    # update positions
    x += v*dt + 0.5*f * dt*dt
    # half update of the velocity
    v += 0.5*f * dt
        
    # compute new forces
    f = compute_forces(x, L)
    # we assume that m=1 for all particles

    # second half update of the velocity
    v += 0.5*f * dt

    return x, v, f

# CONSTANTS
# density
density = 0.7
# number of particles per side
n = 3
# timestep
dt = 0.01
# length of run
tmax = 1.0
# cutoff
rcut = 2.5
# potential shift
shift = -0.016316891136
# total number of particles
N = n*n*n

# particle positions on a cubic lattice
x = empty((3,N))

# TODO: 
# - compute the size of the system
# - set up n*n*n particles on a cubic lattice
# .
# .
# .


# random particle velocities
v = 2.0*random.random((3,N))-1.0

# RUNNING VARIABLES
t = 0.0

# variables to cumulate data
ts = []
Es = []
traj = []

# open the trajectory file
vtffile = open('ljfluid.vtf', 'w')
# write the structure of the system into the file: 
# N particles ("atoms") with a radius of 0.5
vtffile.write('atom 0:%s radius 0.5\n' % (N-1))
vtffile.write('pbc %s %s %s\n' % (L, L, L))

# main loop
f = compute_forces(x, L)
while t < tmax:
    x, v, f = step_vv(x, v, f, dt)
    t += dt

    E = compute_energy(x, v, L)
    print "t=%s, E=%s" % (t, E)

    ts.append(t)
    Es.append(E)
    
    # write out that a new timestep starts
    vtffile.write('timestep\n')
    # write out the coordinates of the particles
    for i in range(N):
        vtffile.write("%s %s %s\n" % (x[0,i], x[1,i], x[2,i]))

vtffile.close()
plot(ts, Es)
        

File: SimPhys_W2/templates/ljfluid.cpp
#include <stdlib.h>
#include <math.h>
#include <iostream>
#include <fstream>

using namespace std;

int N;
double L;
double rcut;
double shift;

// compute the minimum image of particle i and j
double minimum_image(double *x, int i, int j, double rij[3]) {
  rij[0] = x[j] - x[i];
  rij[1] = x[j+N] - x[i+N];
  rij[2] = x[j+2*N] - x[i+2*N];
  
  rij[0] -= rint(rij[0]/L)*L;
  rij[1] -= rint(rij[1]/L)*L;
  rij[2] -= rint(rij[2]/L)*L;
}

void compute_lj_force(double rij[3], double fij[3]) {
  double r2 = rij[0]*rij[0] + rij[1]*rij[1] + rij[2]*rij[2];
  if (r2 < rcut*rcut) {
    double fac = 4.0 * (12.0 * pow(r2, -6.5) - 6.0 * pow(r2, -3.5)) / sqrt(r2);
    fij[0] = fac*rij[0];
    fij[1] = fac*rij[1];
    fij[2] = fac*rij[2];
  } else {
    fij[0] = 0.0;
    fij[1] = 0.0;
    fij[2] = 0.0;
  }
}

double compute_lj_potential(double rij[3]) {
  double r2 = rij[0]*rij[0] + rij[1]*rij[1] + rij[2]*rij[2];
  if (r2 < rcut*rcut) {
    return 4.0 * (pow(r2, -6) - pow(r2, -3)) - shift;
  } else {
    return 0.0;
  }
}

void compute_forces(double *x, double *f) {
  double rij[3];
  double fij[3];

  // set all forces to zero
  for (int i = 0; i < 3*N; i++)
    f[i] = 0.0;

  // add up the forces
  for (int i = 1; i < N; i++) {
    for (int j = 0; j < i; j++) {
      minimum_image(x, i, j, rij);
      compute_lj_force(rij, fij);

      f[i] -= fij[0];
      f[i+N] -= fij[1];
      f[i+2*N] -= fij[2];

      f[j] += fij[0];
      f[j+N] += fij[1];
      f[j+2*N] += fij[2];
    }
  }
}

/*
double compute_energy(double *x, double *v) {
  // TODO:
  // .
  // .
  // .
}
*/

void step_vv(double *x, double *v, double *f, double dt) {
  for (int i = 0; i < N; i++) {
    x[i] += v[i]*dt + 0.5*f[i] * dt*dt;
    x[i+N] += v[i+N]*dt + 0.5*f[i+N] * dt*dt;
    x[i+2*N] += v[i+2*N]*dt + 0.5*f[i+2*N] * dt*dt;
  }

  for (int i = 0; i < N; i++) {
    v[i] += 0.5*f[i] * dt;
    v[i+N] += 0.5*f[i+N] * dt;
    v[i+2*N] += 0.5*f[i+2*N] * dt;
  }

  compute_forces(x, f);

  for (int i = 0; i < N; i++) {
    v[i] += 0.5*f[i] * dt;
    v[i+N] += 0.5*f[i+N] * dt;
    v[i+2*N] += 0.5*f[i+2*N] * dt;
  }
}

int main() {
  // Constants
  const double density = 0.7;
  const int n = 3;
  const double dt = 0.01;
  const double tmax = 1.0;
  rcut = 2.5;
  shift = -0.016316891136;

  N = n*n*n;
  const double volume = N/density;
  L = pow(volume , 1./3.);
  const double l = L/n;

  double x[3*N];
  double v[3*N];
  double f[3*N];

  double t = 0.0;

  /* Set up cubic lattice */
  for (int i = 0; i < n; i++) {
    double x0 = i*l;
    for (int j = 0; j < n; j++) {
      double x1 = j*l;
      for (int k = 0; k < n; k++) {
        double x2 = k*l;
        int ix = i*n*n + j*n + k;
        x[ix] = x0;
        x[ix+N] = x1;
        x[ix+2*N] = x2;

        v[ix] = drand48()*2.0 - 1.0;
        v[ix+N] = drand48()*2.0 - 1.0;
        v[ix+2*N] = drand48()*2.0 - 1.0;
      }
    }
  }
  
  compute_forces(x, f);

  ofstream vtffile("ljfluid_c.vtf");
  vtffile << "atom 0:" << (N-1) << " radius 0.5" << endl;
  vtffile << "unitcell " << L << " " << L << " " << L << endl;

  // ofstream datfile("ljfluid_c.dat");

  while (t < tmax) {
    step_vv(x, v, f, dt);
    t += dt;

    // TODO: implement compute_energy!
    //    double E = compute_energy(x, v);
    // datfile << t << '\t' << E << endl;
    cout << t << endl;

    vtffile << "timestep" << endl;
    for (int i = 0; i < N; i++) {
      vtffile << x[i] << '\t' 
              << x[i+N] << '\t' 
              << x[i+2*N] << endl;
    }
  }

  // datfile.close();
  vtffile.close();

}
